File: src/main/java/in/costea/wiles/data/Token.java
package in.costea.wiles.data;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

public record Token(@NotNull String content, TokenLocation location)
{

    public Token(String content, int line, int lineIndex)
    {
        this(content, new TokenLocation(line, lineIndex));
    }

    public Token(String content)
    {
        this(content, null);
    }

    @Contract(pure = true)
    @Override
    public @NotNull String toString()
    {
        if (location == null)
            return content;
        return "%s at line: %d index: %d".formatted(content, location().line(), location().lineIndex());
    }

    @Contract(value = "null -> false", pure = true)
    @Override
    public boolean equals(Object obj)
    {
        if (obj instanceof Token tokenObj)
        {
            if (tokenObj.location() == null)
                return Objects.equals(content, tokenObj.content);
            return Objects.equals(content, tokenObj.content) && Objects.equals(location(), tokenObj.location());
        }
        return false;
    }
}

File: src/main/java/in/costea/wiles/data/Token.kt
package `in`.costea.wiles.data

import `in`.costea.wiles.constants.ErrorMessages.AT_LINE_INDEX

data class Token(val content: String, val location: TokenLocation) {

    override fun toString(): String {
        return AT_LINE_INDEX.format(content,location.line,location.lineIndex)
    }

    override fun hashCode(): Int {
        var result = content.hashCode()
        result = 31 * result + location.hashCode()
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Token
        if (content != other.content) return false
        if (location != other.location) return false
        return true
    }
}
